Assets/Features/Building/Scripts/Grid/Grid.cs

```diff
@@ -0,0 +1,391 @@
+        /// y = y-axis
+        /// z = z-axis for multiple layers for example floor tile layer and decoration layer
+        /// </summary>
+        private CellData[,,] _cells;
+
+        [SerializeField] private BuildableAtlas _atlas;
+        [SerializeField] private Tilemap _tilemap;
+
+        [SerializeField] private Vector3Int _gridSize;
+        [SerializeField] private float _cellSize;
+
+        [SerializeField] private List<List<Vector3Int>> _cellGroup;
+
+        /// <summary>
+        /// If true the map will be updated at the end of the frame and set to false
+        /// </summary>
+        private bool _mapUpdated;
+
+
+        public Vector3Int GridSize => _gridSize;
+        public float CellSize => _cellSize;
+
+        void Start()
+        {
+            if (_atlas == null)
+                Debug.LogError("No atlas is assigned!");
+
+            _cells = new CellData[_gridSize.x, _gridSize.y, _gridSize.z];
+            _cellGroup = new List<List<Vector3Int>>();
+            PopulateCells();
+        }
+
+        /// <summary>
+        /// Creates a flattend 2d array to see if it is traversable
+        /// </summary>
+        /// <returns>A flattend 2d bool array with false as traversable</returns>
+        public bool[,] UnTraversable()
+        {
+            bool[,] unTraversable = new bool[_gridSize.x, _gridSize.y];
+
+            for (int z = 0; z < _gridSize.z; z++)
+            {
+                for (int x = 0; x < _gridSize.x; x++)
+                {
+                    for (int y = 0; y < _gridSize.y; y++)
+                    {
+                        if (unTraversable[x, y] || _cells[x, y, z].Tile == CellData.empty.Tile)
+                            continue;
+
+                        if (_cells[x, y, z].BuildPercentage == 1f)
+                            unTraversable[x, y] = _atlas.Items[_cells[x, y, z].Tile].UnTraversable;
+                    }
+                }
+            }
+
+            return unTraversable;
+        }
+
+        /// <summary>
+        /// Sets the default value of the whole array
+        /// </summary>
+        private void PopulateCells()
+        {
+            for (int x = 0; x < _gridSize.x; x++)
+            {
+                for (int y = 0; y < _gridSize.y; y++)
+                {
+                    for (int z = 0; z < _gridSize.z; z++)
+                    {
+                        _cells[x, y, z] = CellData.empty;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// If the map update was called this frame update it
+        /// </summary>
+        private void LateUpdate()
+        {
+            if (_mapUpdated)
+            {
+                UpdateMap();
+                _mapUpdated = false;
+            }
+        }
+
+        /// <summary>
+        /// Loops through the whole grid and sets all the cells to what they are supposed to be
+        /// TODO add a buffer, so correct tiles aren't changed
```

Comment:
Put this in he backlog if necessary, but remove the comment since it will get lost in the big codebase.
I know you can look for TODO tags, but keeping this in the backlog is easier